[Bug] [mongodb] pull big collection (>20 billion records) lead to checkpoint hang or OOM

### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.


### Flink version

1.17.0

### Flink CDC version

2.4.1

### Database and its version

Using MongoDB:          4.2.11

### Minimal reproduce step

1. prepare a mongodb collection with more than 10 billion records
2. pull records from mongodb, sync to paimon(parquet format on hdfs) or map record to integer 1 and set  up a minute window to count. 
3.  when the application runs for some time (30 minutes ~ a few hours)，it hangs or run into OOM exception and can recover.



### What did you expect to see?

the application runs without error and pull all records from mongodb.

### What did you see instead?

the application runs for server minutes and fail, only 5% less records are pulled from mongodb.

### Anything else?

this is the metrics without sink, only count pulled records.

![cdc_topo](https://github.com/ververica/flink-cdc-connectors/assets/64360028/2e3e0a73-4744-4adb-bda7-4b4c25d24b39)


![cdc_checkpoint](https://github.com/ververica/flink-cdc-connectors/assets/64360028/4ef7b165-bf87-471c-8b91-5c6f99fb6d57)


![cdc_metric1](https://github.com/ververica/flink-cdc-connectors/assets/64360028/a13b0e43-d045-40dd-a047-6989daf92bb6)

![cdc_metric2](https://github.com/ververica/flink-cdc-connectors/assets/64360028/bd32083f-639b-4a2a-b9bc-2276c64d93e6)



### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

2023-09-13 13:53:19.930 [SourceCoordinator-Source: 100205_mongocdc_t_lint_defect_v2_prod] INFO  c.v.c.c.base.source.enumerator.IncrementalSourceEnumerator  - Assign split SnapshotSplit{tableId=db_defect.t_lint_defect_v2, splitId='db_defect.t_lint_defect_v2:926', splitKeyType=[`task_id` INT], splitStart=[{"task_id": "hashed"}, {"task_id": -8923183964006431007}], splitEnd=[{"task_id": "hashed"}, {"task_id": -8922499848073591815}], highWatermark=null} to subtask 16
2023-09-13 13:53:19.931 [SourceCoordinator-Source: 100205_mongocdc_t_lint_defect_v2_prod] INFO  c.v.c.c.base.source.enumerator.IncrementalSourceEnumerator  - The enumerator receives finished split offsets FinishedSnapshotSplitsReportEvent{finishedOffsets={db_defect.t_lint_defect_v2:917={resumeToken={"_data": "8265014DFD000012582B0229296E04"}, timestamp=7278184221829501528}}} from subtask 16.
2023-09-13 13:53:23.002 [flink-akka.actor.default-dispatcher-22] INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph  - Writer : mongocdc_t_lint_defect_v2_prod_100205 (13/32) (64269a3ffd76aa7f28c190300cb2c08c_9dd63673dd41ea021b896d5203f3ba7c_12_4) switched from RUNNING to FAILED on cdc-5766833c4d1e47a6a8d49d9599e20d2f-taskmanager-1-12 @ xx.xx.xx.xx (dataPort=43393).
org.apache.flink.runtime.io.network.netty.exception.LocalTransportException: GC overhead limit exceeded (connection to 'xx.xx.xx.xx/xx.xx.xx.xx:34779 [cdc-5766833c4d1e47a6a8d49d9599e20d2f-taskmanager-1-18]')
        at org.apache.flink.runtime.io.network.netty.CreditBasedPartitionRequestClientHandler.exceptionCaught(CreditBasedPartitionRequestClientHandler.java:175)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:302)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:281)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:273)
        at org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:143)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:302)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:381)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at org.apache.flink.shaded.netty4.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at org.apache.flink.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at org.apache.flink.shaded.netty4.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at org.apache.flink.shaded.netty4.io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)
        at org.apache.flink.shaded.netty4.io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499)
        at org.apache.flink.shaded.netty4.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:397)
        at org.apache.flink.shaded.netty4.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at org.apache.flink.shaded.netty4.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
2023-09-13 13:53:23.003 [flink-akka.actor.default-dispatcher-22] INFO  org.apache.flink.runtime.jobmaster.JobMaster  - 65 tasks will be restarted to recover the failed task 64269a3ffd76aa7f28c190300cb2c08c_9dd63673dd41ea021b896d5203f3ba7c_12_4.
2023-09-13 13:53:23.003 [flink-akka.actor.default-dispatcher-22] INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph  - Job mongodb CDC to paimon for 100205_mongocdc_t_lint_defect_v2_prod (ffffffffff2fce170000000000000000) switched from state RUNNING to RESTARTING.
2023-09-13 13:53:23.003 [flink-akka.actor.default-dispatcher-22] WARN  org.apache.flink.runtime.checkpoint.CheckpointFailureManager  - Failed to trigger or complete checkpoint 15 for job ffffffffff2fce170000000000000000. (0 consecutive failed attempts so far)
org.apache.flink.runtime.checkpoint.CheckpointException: Checkpoint Coordinator is suspending.
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.stopCheckpointScheduler(CheckpointCoordinator.java:1977)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinatorDeActivator.jobStatusChanges(CheckpointCoordinatorDeActivator.java:46)
        at org.apache.flink.runtime.executiongraph.DefaultExecutionGraph.notifyJobStatusChange(DefaultExecutionGraph.java:1608)
        at org.apache.flink.runtime.executiongraph.DefaultExecutionGraph.transitionState(DefaultExecutionGraph.java:1174)
        at org.apache.flink.runtime.executiongraph.DefaultExecutionGraph.transitionState(DefaultExecutionGraph.java:1146)
        at org.apache.flink.runtime.scheduler.SchedulerBase.transitionExecutionGraphState(SchedulerBase.java:581)
        at org.apache.flink.runtime.scheduler.DefaultScheduler.addVerticesToRestartPending(DefaultScheduler.java:364)
        at org.apache.flink.runtime.scheduler.DefaultScheduler.restartTasksWithDelay(DefaultScheduler.java:336)
        at org.apache.flink.runtime.scheduler.DefaultScheduler.maybeRestartTasks(DefaultScheduler.java:307)
        at org.apache.flink.runtime.scheduler.DefaultScheduler.handleTaskFailure(DefaultScheduler.java:249)
        at org.apache.flink.runtime.scheduler.DefaultScheduler.onTaskFailed(DefaultScheduler.java:242)
        at org.apache.flink.runtime.scheduler.SchedulerBase.onTaskExecutionStateUpdate(SchedulerBase.java:748)
        at org.apache.flink.runtime.scheduler.SchedulerBase.updateTaskExecutionState(SchedulerBase.java:725)
        at org.apache.flink.runtime.scheduler.SchedulerNG.updateTaskExecutionState(SchedulerNG.java:80)
        at org.apache.flink.runtime.jobmaster.JobMaster.updateTaskExecutionState(JobMaster.java:479)
        at sun.reflect.GeneratedMethodAccessor434.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.lambda$handleRpcInvocation$1(AkkaRpcActor.java:309)
        at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:83)
        at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcInvocation(AkkaRpcActor.java:307)
        at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcMessage(AkkaRpcActor.java:222)

Try to split the chunk into small pieces, but I found that this is not going to work. In some case, one shard key may contains millions of records. In the log below, there are 3119249 records with task_id equals to 1559426479859094090 (hashed), this will lead to the OOM exception in task manager.

@Jiabao-Sun 

```
2023-10-18 13:58:59.766 [SourceCoordinator-Source: 100205_mongocdc_t_lint_defect_v2_dev2] INFO  c.v.c.c.m.s.assigners.splitters.SplitVectorSplitStrategy  - Collection {"task_id": 1559426479859094090} {"task_id": 1559426479859094091} - db_defect.t_lint_defect_v2: dataSize cmd result {"size": 3.268972952E9, "numObjects": 3119249.0, "millis": 1718, "ok": 1.0, "operationTime": {"$timestamp": {"t": 1697608732, "i": 1}}, "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1697608736, "i": 4}}, "signature": {"hash": {"$binary": {"base64": "lgfHxhHe0+DNMa16e4a7wv/OFAU=", "subType": "00"}}, "keyId": 7248542109899161603}}}
```

```
Caused by: org.apache.flink.util.FlinkRuntimeException: Read split SnapshotSplit{tableId=db_defect.t_lint_defect_v2, splitId='db_defect.t_lint_defect_v2:1090', splitKeyType=[`task_id` INT], splitStart=[{"task_id": "hashed"}, {"task_id": 1559426479859094090}], splitEnd=[{"task_id": "hashed"}, {"task_id": 1559426479859094091}], highWatermark=null} error due to GC overhead limit exceeded.
	at com.ververica.cdc.connectors.base.source.reader.external.IncrementalSourceScanFetcher.checkReadException(IncrementalSourceScanFetcher.java:179)
	at com.ververica.cdc.connectors.base.source.reader.external.IncrementalSourceScanFetcher.pollSplitRecords(IncrementalSourceScanFetcher.java:126)
	at com.ververica.cdc.connectors.base.source.reader.IncrementalSourceSplitReader.fetch(IncrementalSourceSplitReader.java:73)
	at org.apache.flink.connector.base.source.reader.fetcher.FetchTask.run(FetchTask.java:58)
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:162)
	... 6 more
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.util.Arrays.copyOf(Arrays.java:3332)
	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
```

--- a/docs/content/connectors/mongodb-cdc(ZH).md
+++ b/docs/content/connectors/mongodb-cdc(ZH).md
@@ -273,7 +273,7 @@ upstart 流需要一个唯一的密钥，所以我们必须声明 `_id` 作为�
     <tr>
       <td>scan.incremental.snapshot.chunk.samples</td>
       <td>optional</td>
-      <td style="word-wrap: break-word;">64</td>
+      <td style="word-wrap: break-word;">20</td>
       <td>Integer</td>
       <td>读取增量快照时，如果采用采样方式划分chunk，每个chunk采样的数据条数。</td>
     </tr>

--- a/docs/content/connectors/mongodb-cdc(ZH).md
+++ b/docs/content/connectors/mongodb-cdc(ZH).md
@@ -271,6 +271,13 @@ upstart 流需要一个唯一的密钥，所以我们必须声明 `_id` 作为�
       <td>增量快照的区块大小 mb。</td>
     </tr>
     <tr>
+      <td>scan.incremental.snapshot.chunk.samples</td>
+      <td>optional</td>
+      <td style="word-wrap: break-word;">64</td>
+      <td>Integer</td>
+      <td>读取增量快照时，如果采用采样方式划分chunk，每个chunk采样的数据条数。</td>
+    </tr>
+    <tr>
       <td>scan.incremental.close-idle-reader.enabled</td>
       <td>optional</td>
       <td style="word-wrap: break-word;">false</td>

--- a/docs/content/connectors/mongodb-cdc(ZH).md
+++ b/docs/content/connectors/mongodb-cdc(ZH).md
@@ -275,7 +275,7 @@ upstart 流需要一个唯一的密钥，所以我们必须声明 `_id` 作为�
       <td>optional</td>
       <td style="word-wrap: break-word;">20</td>
       <td>Integer</td>
-      <td>读取增量快照时，如果采用采样方式划分chunk，每个chunk采样的数据条数。</td>
+      <td>采样分片策略，每个chunk采样的数据条数。</td>
     </tr>
     <tr>
       <td>scan.incremental.close-idle-reader.enabled</td>

--- a/docs/content/connectors/mongodb-cdc.md
+++ b/docs/content/connectors/mongodb-cdc.md
@@ -277,6 +277,13 @@ Connector Options
       <td>The chunk size mb of incremental snapshot.</td>
     </tr>
     <tr>
+      <td>scan.incremental.snapshot.chunk.samples</td>
+      <td>optional</td>
+      <td style="word-wrap: break-word;">64</td>
+      <td>Integer</td>
+      <td>The sample count per chunk when using SampleBucketSplitStrategy during incremental snapshot.</td>
+    </tr>
+    <tr>
       <td>scan.incremental.close-idle-reader.enabled</td>
       <td>optional</td>
       <td style="word-wrap: break-word;">false</td>

--- a/docs/content/connectors/mongodb-cdc.md
+++ b/docs/content/connectors/mongodb-cdc.md
@@ -279,7 +279,7 @@ Connector Options
     <tr>
       <td>scan.incremental.snapshot.chunk.samples</td>
       <td>optional</td>
-      <td style="word-wrap: break-word;">64</td>
+      <td style="word-wrap: break-word;">20</td>
       <td>Integer</td>
       <td>The sample count per chunk when using SampleBucketSplitStrategy during incremental snapshot.</td>
     </tr>

--- a/docs/content/connectors/mongodb-cdc.md
+++ b/docs/content/connectors/mongodb-cdc.md
@@ -281,7 +281,7 @@ Connector Options
       <td>optional</td>
       <td style="word-wrap: break-word;">20</td>
       <td>Integer</td>
-      <td>The sample count per chunk when using SampleBucketSplitStrategy during incremental snapshot.</td>
+      <td>The samples count per chunk when using sample partition strategy during incremental snapshot.</td>
     </tr>
     <tr>
       <td>scan.incremental.close-idle-reader.enabled</td>

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/MongoDBSourceBuilder.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/MongoDBSourceBuilder.java
@@ -179,6 +179,16 @@ public class MongoDBSourceBuilder<T> {
     }
 
     /**
+     * scan.incremental.snapshot.chunk.samples
+     *
+     * <p>The number of samples to take per chunk. Defaults to 20.
+     */
+    public MongoDBSourceBuilder<T> samplesPerChunk(int samplesPerChunk) {
+        this.configFactory.samplesPerChunk(samplesPerChunk);
+        return this;
+    }
+
+    /**
      * scan.incremental.close-idle-reader.enabled
      *
      * <p>Whether to close idle readers at the end of the snapshot phase. This feature depends on

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SampleBucketSplitStrategy.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SampleBucketSplitStrategy.java
@@ -26,6 +26,8 @@ import io.debezium.relational.history.TableChanges;
 import org.bson.BsonDocument;
 import org.bson.BsonValue;
 import org.bson.conversions.Bson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -66,9 +68,13 @@ import static com.ververica.cdc.connectors.mongodb.source.utils.MongoUtils.colle
 @Internal
 public class SampleBucketSplitStrategy implements SplitStrategy {
 
+    private static final Logger LOG = LoggerFactory.getLogger(SampleBucketSplitStrategy.class);
+
     public static final SampleBucketSplitStrategy INSTANCE = new SampleBucketSplitStrategy();
     private static final int DEFAULT_SAMPLING_THRESHOLD = 102400;
     private static final double DEFAULT_SAMPLING_RATE = 0.05;
+    private static final int SAMPLING_COUNT_BY_RATE_THRESHOLD = 1000000;
+    private static final int MAX_SAMPLE_PER_CHUNK = 20;
 
     private SampleBucketSplitStrategy() {}
 
@@ -93,6 +99,10 @@ public class SampleBucketSplitStrategy implements SplitStrategy {
         } else {
             // sampled using sample rate.
             numberOfSamples = (int) Math.floor(count * DEFAULT_SAMPLING_RATE);
+            // avoid sampling too much records on a huge collection
+            if (numberOfSamples > SAMPLING_COUNT_BY_RATE_THRESHOLD) {
+                numberOfSamples = numChunks * MAX_SAMPLE_PER_CHUNK;
+            }
         }
 
         TableId collectionId = splitContext.getCollectionId();
@@ -105,9 +115,18 @@ public class SampleBucketSplitStrategy implements SplitStrategy {
             pipeline.add(sample(numberOfSamples));
         }
         pipeline.add(bucketAuto("$" + ID_FIELD, numChunks));
+        LOG.info(
+                "Collection {} going to sample {} records into {} chunks",
+                collectionId,
+                numberOfSamples,
+                numChunks);
 
         List<BsonDocument> chunks =
                 collection.aggregate(pipeline).allowDiskUse(true).into(new ArrayList<>());
+        LOG.info(
+                "Collection {} got {} chunks by auto bucket and sample",
+                collectionId,
+                chunks.size());
 
         RowType rowType = shardKeysToRowType(Collections.singleton(ID_FIELD));
 

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SampleBucketSplitStrategy.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SampleBucketSplitStrategy.java
@@ -72,9 +72,6 @@ public class SampleBucketSplitStrategy implements SplitStrategy {
 
     public static final SampleBucketSplitStrategy INSTANCE = new SampleBucketSplitStrategy();
     private static final int DEFAULT_SAMPLING_THRESHOLD = 102400;
-    private static final double DEFAULT_SAMPLING_RATE = 0.05;
-    private static final int SAMPLING_COUNT_BY_RATE_THRESHOLD = 1000000;
-    private static final int MAX_SAMPLE_PER_CHUNK = 20;
 
     private SampleBucketSplitStrategy() {}
 
@@ -97,12 +94,8 @@ public class SampleBucketSplitStrategy implements SplitStrategy {
             // full sampling if document count less than sampling size threshold.
             numberOfSamples = (int) count;
         } else {
-            // sampled using sample rate.
-            numberOfSamples = (int) Math.floor(count * DEFAULT_SAMPLING_RATE);
             // avoid sampling too much records on a huge collection
-            if (numberOfSamples > SAMPLING_COUNT_BY_RATE_THRESHOLD) {
-                numberOfSamples = numChunks * MAX_SAMPLE_PER_CHUNK;
-            }
+            numberOfSamples = numChunks * splitContext.getSamplesPerChunk();
         }
 
         TableId collectionId = splitContext.getCollectionId();

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SampleBucketSplitStrategy.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SampleBucketSplitStrategy.java
@@ -94,8 +94,7 @@ public class SampleBucketSplitStrategy implements SplitStrategy {
             // full sampling if document count less than sampling size threshold.
             numberOfSamples = (int) count;
         } else {
-            // avoid sampling too much records on a huge collection
-            numberOfSamples = numChunks * splitContext.getSamplesPerChunk();
+            numberOfSamples = Math.min(numChunks * splitContext.getSamplesPerChunk(), (int) count);
         }
 
         TableId collectionId = splitContext.getCollectionId();

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/ShardedSplitStrategy.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/ShardedSplitStrategy.java
@@ -60,7 +60,7 @@ import static com.ververica.cdc.connectors.mongodb.source.utils.MongoUtils.readC
 @Internal
 public class ShardedSplitStrategy implements SplitStrategy {
 
-    private static final Logger LOG = LoggerFactory.getLogger(SplitVectorSplitStrategy.class);
+    private static final Logger LOG = LoggerFactory.getLogger(ShardedSplitStrategy.class);
 
     public static final ShardedSplitStrategy INSTANCE = new ShardedSplitStrategy();
 
@@ -75,9 +75,10 @@ public class ShardedSplitStrategy implements SplitStrategy {
         BsonDocument collectionMetadata;
         try {
             collectionMetadata = readCollectionMetadata(mongoClient, collectionId);
-            if (!isValidShardedCollection(collectionMetadata)) {
+            if (!(isValidShardedCollection(collectionMetadata)
+                    && isNotShardedByHash(collectionMetadata))) {
                 LOG.warn(
-                        "Collection {} does not appear to be sharded, fallback to SampleSplitter.",
+                        "Collection {} does not appear to be sharded, or shared by hash, fallback to SampleSplitter.",
                         collectionId);
                 return SampleBucketSplitStrategy.INSTANCE.split(splitContext);
             }
@@ -128,5 +129,11 @@ public class ShardedSplitStrategy implements SplitStrategy {
     private boolean isValidShardedCollection(BsonDocument collectionMetadata) {
         return collectionMetadata != null
                 && !collectionMetadata.getBoolean(DROPPED_FIELD, BsonBoolean.FALSE).getValue();
+    }
+
+    private boolean isNotShardedByHash(BsonDocument collectionMetadata) {
+        BsonDocument splitKeys = collectionMetadata.getDocument(KEY_FIELD);
+        return splitKeys.values().stream()
+                .noneMatch(v -> v.isString() && v.asString().getValue().equals("hashed"));
     }
 }

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SplitContext.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/assigners/splitters/SplitContext.java
@@ -39,16 +39,19 @@ public class SplitContext {
     private final TableId collectionId;
     private final BsonDocument collectionStats;
     private final int chunkSizeMB;
+    private final int samplesPerChunk;
 
     public SplitContext(
             MongoClient mongoClient,
             TableId collectionId,
             BsonDocument collectionStats,
-            int chunkSizeMB) {
+            int chunkSizeMB,
+            int samplesPerChunk) {
         this.mongoClient = mongoClient;
         this.collectionId = collectionId;
         this.collectionStats = collectionStats;
         this.chunkSizeMB = chunkSizeMB;
+        this.samplesPerChunk = samplesPerChunk;
     }
 
     public static SplitContext of(MongoDBSourceConfig sourceConfig, TableId collectionId) {
@@ -57,7 +60,8 @@ public class SplitContext {
                 mongoClient,
                 collectionId,
                 collStats(mongoClient, collectionId),
-                sourceConfig.getSplitSize());
+                sourceConfig.getSplitSize(),
+                sourceConfig.getSamplesPerChunk());
     }
 
     public MongoClient getMongoClient() {
@@ -70,6 +74,10 @@ public class SplitContext {
 
     public int getChunkSizeMB() {
         return chunkSizeMB;
+    }
+
+    public int getSamplesPerChunk() {
+        return samplesPerChunk;
     }
 
     /** The number of objects or documents in this collection. */

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceConfig.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceConfig.java
@@ -48,6 +48,7 @@ public class MongoDBSourceConfig implements SourceConfig {
     private final int heartbeatIntervalMillis;
     private final int splitMetaGroupSize;
     private final int splitSizeMB;
+    private final int samplesPerChunk;
     private final boolean closeIdleReaders;
     private final boolean enableFullDocPrePostImage;
     private final boolean disableCursorTimeout;
@@ -68,6 +69,7 @@ public class MongoDBSourceConfig implements SourceConfig {
             int heartbeatIntervalMillis,
             int splitMetaGroupSize,
             int splitSizeMB,
+            int samplesPerChunk,
             boolean closeIdleReaders,
             boolean enableFullDocPrePostImage,
             boolean disableCursorTimeout) {
@@ -87,6 +89,7 @@ public class MongoDBSourceConfig implements SourceConfig {
         this.heartbeatIntervalMillis = heartbeatIntervalMillis;
         this.splitMetaGroupSize = splitMetaGroupSize;
         this.splitSizeMB = splitSizeMB;
+        this.samplesPerChunk = samplesPerChunk;
         this.closeIdleReaders = closeIdleReaders;
         this.enableFullDocPrePostImage = enableFullDocPrePostImage;
         this.disableCursorTimeout = disableCursorTimeout;
@@ -159,6 +162,10 @@ public class MongoDBSourceConfig implements SourceConfig {
         return splitMetaGroupSize;
     }
 
+    public int getSamplesPerChunk() {
+        return samplesPerChunk;
+    }
+
     @Override
     public boolean isIncludeSchemaChanges() {
         return false;
@@ -194,6 +201,7 @@ public class MongoDBSourceConfig implements SourceConfig {
                 && heartbeatIntervalMillis == that.heartbeatIntervalMillis
                 && splitMetaGroupSize == that.splitMetaGroupSize
                 && splitSizeMB == that.splitSizeMB
+                && samplesPerChunk == that.samplesPerChunk
                 && closeIdleReaders == that.closeIdleReaders
                 && Objects.equals(scheme, that.scheme)
                 && Objects.equals(hosts, that.hosts)
@@ -222,6 +230,7 @@ public class MongoDBSourceConfig implements SourceConfig {
                 heartbeatIntervalMillis,
                 splitMetaGroupSize,
                 splitSizeMB,
+                samplesPerChunk,
                 closeIdleReaders);
     }
 }

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceConfigFactory.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceConfigFactory.java
@@ -32,6 +32,7 @@ import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOp
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.HEARTBEAT_INTERVAL_MILLIS;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.POLL_AWAIT_TIME_MILLIS;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.POLL_MAX_BATCH_SIZE;
+import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCHEME;
 import static org.apache.flink.util.Preconditions.checkArgument;
@@ -58,6 +59,7 @@ public class MongoDBSourceConfigFactory implements Factory<MongoDBSourceConfig> 
     private Integer heartbeatIntervalMillis = HEARTBEAT_INTERVAL_MILLIS.defaultValue();
     private Integer splitMetaGroupSize = CHUNK_META_GROUP_SIZE.defaultValue();
     private Integer splitSizeMB = SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB.defaultValue();
+    private Integer samplesPerChunk = SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES.defaultValue();
     private boolean closeIdleReaders = false;
     private boolean enableFullDocPrePostImage = false;
     private boolean disableCursorTimeout = true;
@@ -207,6 +209,17 @@ public class MongoDBSourceConfigFactory implements Factory<MongoDBSourceConfig> 
     }
 
     /**
+     * scan.incremental.snapshot.chunk.samples
+     *
+     * <p>The number of samples to take per chunk. Defaults to 20.
+     */
+    public MongoDBSourceConfigFactory samplesPerChunk(int samplesPerChunk) {
+        checkArgument(samplesPerChunk > 0);
+        this.samplesPerChunk = samplesPerChunk;
+        return this;
+    }
+
+    /**
      * Whether to close idle readers at the end of the snapshot phase. This feature depends on
      * FLIP-147: Support Checkpoints After Tasks Finished. The flink version is required to be
      * greater than or equal to 1.14, and the configuration <code>
@@ -261,6 +274,7 @@ public class MongoDBSourceConfigFactory implements Factory<MongoDBSourceConfig> 
                 heartbeatIntervalMillis,
                 splitMetaGroupSize,
                 splitSizeMB,
+                samplesPerChunk,
                 closeIdleReaders,
                 enableFullDocPrePostImage,
                 disableCursorTimeout);

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/source/config/MongoDBSourceOptions.java
@@ -135,6 +135,13 @@ public class MongoDBSourceOptions {
                             "The chunk size mb of incremental snapshot. Defaults to 64mb.");
 
     @Experimental
+    public static final ConfigOption<Integer> SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES =
+            ConfigOptions.key("scan.incremental.snapshot.chunk.samples")
+                    .intType()
+                    .defaultValue(20)
+                    .withDescription("The number of samples to take per chunk. Defaults to 20.");
+
+    @Experimental
     public static final ConfigOption<Boolean> FULL_DOCUMENT_PRE_POST_IMAGE =
             ConfigOptions.key("scan.full-changelog")
                     .booleanType()

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSource.java
@@ -79,6 +79,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
     private final boolean enableParallelRead;
     private final Integer splitMetaGroupSize;
     private final Integer splitSizeMB;
+    private final Integer samplesPerChunk;
     private final boolean closeIdlerReaders;
     private final boolean enableFullDocPrePostImage;
     private final boolean noCursorTimeout;
@@ -112,6 +113,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
             boolean enableParallelRead,
             @Nullable Integer splitMetaGroupSize,
             @Nullable Integer splitSizeMB,
+            @Nullable Integer samplesPerChunk,
             boolean closeIdlerReaders,
             boolean enableFullDocPrePostImage,
             boolean noCursorTimeout) {
@@ -135,6 +137,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
         this.enableParallelRead = enableParallelRead;
         this.splitMetaGroupSize = splitMetaGroupSize;
         this.splitSizeMB = splitSizeMB;
+        this.samplesPerChunk = samplesPerChunk;
         this.closeIdlerReaders = closeIdlerReaders;
         this.enableFullDocPrePostImage = enableFullDocPrePostImage;
         this.noCursorTimeout = noCursorTimeout;
@@ -209,6 +212,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
                     .ifPresent(builder::heartbeatIntervalMillis);
             Optional.ofNullable(splitMetaGroupSize).ifPresent(builder::splitMetaGroupSize);
             Optional.ofNullable(splitSizeMB).ifPresent(builder::splitSizeMB);
+            Optional.ofNullable(samplesPerChunk).ifPresent(builder::samplesPerChunk);
             return SourceProvider.of(builder.build());
         } else {
             com.ververica.cdc.connectors.mongodb.MongoDBSource.Builder<RowData> builder =
@@ -289,6 +293,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
                         enableParallelRead,
                         splitMetaGroupSize,
                         splitSizeMB,
+                        samplesPerChunk,
                         closeIdlerReaders,
                         enableFullDocPrePostImage,
                         noCursorTimeout);
@@ -324,6 +329,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
                 && Objects.equals(enableParallelRead, that.enableParallelRead)
                 && Objects.equals(splitMetaGroupSize, that.splitMetaGroupSize)
                 && Objects.equals(splitSizeMB, that.splitSizeMB)
+                && Objects.equals(samplesPerChunk, that.samplesPerChunk)
                 && Objects.equals(producedDataType, that.producedDataType)
                 && Objects.equals(metadataKeys, that.metadataKeys)
                 && Objects.equals(closeIdlerReaders, that.closeIdlerReaders)
@@ -352,6 +358,7 @@ public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetad
                 enableParallelRead,
                 splitMetaGroupSize,
                 splitSizeMB,
+                samplesPerChunk,
                 producedDataType,
                 metadataKeys,
                 closeIdlerReaders,

--- a/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSourceFactory.java
+++ b/flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableSourceFactory.java
@@ -49,6 +49,7 @@ import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOp
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.PASSWORD;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.POLL_AWAIT_TIME_MILLIS;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.POLL_MAX_BATCH_SIZE;
+import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_ENABLED;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_NO_CURSOR_TIMEOUT;
@@ -104,6 +105,7 @@ public class MongoDBTableSourceFactory implements DynamicTableSourceFactory {
 
         int splitSizeMB = config.get(SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB);
         int splitMetaGroupSize = config.get(CHUNK_META_GROUP_SIZE);
+        int samplesPerChunk = config.get(SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES);
 
         boolean enableFullDocumentPrePostImage =
                 config.getOptional(FULL_DOCUMENT_PRE_POST_IMAGE).orElse(false);
@@ -135,6 +137,7 @@ public class MongoDBTableSourceFactory implements DynamicTableSourceFactory {
                 enableParallelRead,
                 splitMetaGroupSize,
                 splitSizeMB,
+                samplesPerChunk,
                 enableCloseIdleReaders,
                 enableFullDocumentPrePostImage,
                 noCursorTimeout);
@@ -207,6 +210,7 @@ public class MongoDBTableSourceFactory implements DynamicTableSourceFactory {
         options.add(HEARTBEAT_INTERVAL_MILLIS);
         options.add(SCAN_INCREMENTAL_SNAPSHOT_ENABLED);
         options.add(SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB);
+        options.add(SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES);
         options.add(CHUNK_META_GROUP_SIZE);
         options.add(SCAN_INCREMENTAL_CLOSE_IDLE_READER_ENABLED);
         options.add(FULL_DOCUMENT_PRE_POST_IMAGE);

--- a/flink-connector-mongodb-cdc/src/test/java/com/ververica/cdc/connectors/mongodb/source/MongoDBFullChangelogITCase.java
+++ b/flink-connector-mongodb-cdc/src/test/java/com/ververica/cdc/connectors/mongodb/source/MongoDBFullChangelogITCase.java
@@ -76,6 +76,7 @@ public class MongoDBFullChangelogITCase extends MongoDBSourceTestBase {
                 new MongoDBSourceConfigFactory()
                         .hosts(CONTAINER.getHostAndPort())
                         .splitSizeMB(1)
+                        .samplesPerChunk(10)
                         .pollAwaitTimeMillis(500)
                         .create(0);
 

--- a/flink-connector-mongodb-cdc/src/test/java/com/ververica/cdc/connectors/mongodb/source/reader/MongoDBSnapshotSplitReaderTest.java
+++ b/flink-connector-mongodb-cdc/src/test/java/com/ververica/cdc/connectors/mongodb/source/reader/MongoDBSnapshotSplitReaderTest.java
@@ -79,6 +79,7 @@ public class MongoDBSnapshotSplitReaderTest extends MongoDBSourceTestBase {
                         .username(FLINK_USER)
                         .password(FLINK_USER_PASSWORD)
                         .splitSizeMB(1)
+                        .samplesPerChunk(10)
                         .pollAwaitTimeMillis(500);
 
         sourceConfig = configFactory.create(0);

--- a/flink-connector-mongodb-cdc/src/test/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableFactoryTest.java
+++ b/flink-connector-mongodb-cdc/src/test/java/com/ververica/cdc/connectors/mongodb/table/MongoDBTableFactoryTest.java
@@ -53,6 +53,7 @@ import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOp
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.HEARTBEAT_INTERVAL_MILLIS;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.POLL_AWAIT_TIME_MILLIS;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.POLL_MAX_BATCH_SIZE;
+import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_ENABLED;
 import static com.ververica.cdc.connectors.mongodb.source.config.MongoDBSourceOptions.SCAN_NO_CURSOR_TIMEOUT;
@@ -104,6 +105,8 @@ public class MongoDBTableFactoryTest {
             SCAN_INCREMENTAL_SNAPSHOT_ENABLED.defaultValue();
     private static final int SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB_DEFAULT =
             SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB.defaultValue();
+    private static final int SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES_DEFAULT =
+            SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES.defaultValue();
     private static final int CHUNK_META_GROUP_SIZE_DEFAULT = CHUNK_META_GROUP_SIZE.defaultValue();
     private static final boolean SCAN_INCREMENTAL_CLOSE_IDLE_READER_ENABLED_DEFAULT =
             SCAN_INCREMENTAL_CLOSE_IDLE_READER_ENABLED.defaultValue();
@@ -140,6 +143,7 @@ public class MongoDBTableFactoryTest {
                         SCAN_INCREMENTAL_SNAPSHOT_ENABLED_DEFAULT,
                         CHUNK_META_GROUP_SIZE_DEFAULT,
                         SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB_DEFAULT,
+                        SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES_DEFAULT,
                         SCAN_INCREMENTAL_CLOSE_IDLE_READER_ENABLED_DEFAULT,
                         FULL_DOCUMENT_PRE_POST_IMAGE_ENABLED_DEFAULT,
                         SCAN_NO_CURSOR_TIMEOUT_DEFAULT);
@@ -161,6 +165,7 @@ public class MongoDBTableFactoryTest {
         options.put("scan.incremental.snapshot.enabled", "true");
         options.put("chunk-meta.group.size", "1001");
         options.put("scan.incremental.snapshot.chunk.size.mb", "10");
+        options.put("scan.incremental.snapshot.chunk.samples", "10");
         options.put("scan.incremental.close-idle-reader.enabled", "true");
         options.put("scan.full-changelog", "true");
         options.put("scan.cursor.no-timeout", "false");
@@ -185,6 +190,7 @@ public class MongoDBTableFactoryTest {
                         LOCAL_TIME_ZONE,
                         true,
                         1001,
+                        10,
                         10,
                         true,
                         true,
@@ -224,6 +230,7 @@ public class MongoDBTableFactoryTest {
                         SCAN_INCREMENTAL_SNAPSHOT_ENABLED_DEFAULT,
                         CHUNK_META_GROUP_SIZE_DEFAULT,
                         SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SIZE_MB_DEFAULT,
+                        SCAN_INCREMENTAL_SNAPSHOT_CHUNK_SAMPLES_DEFAULT,
                         SCAN_INCREMENTAL_CLOSE_IDLE_READER_ENABLED_DEFAULT,
                         FULL_DOCUMENT_PRE_POST_IMAGE_ENABLED_DEFAULT,
                         SCAN_NO_CURSOR_TIMEOUT_DEFAULT);

